zbus: add zbus_chan_publisher for scheduling messages

brandon-exact opened the pull request:
this PR introduces the concept of zbus publishers, a async method of calling zbus_chan_pub() from a work queue at a scheduled time. under the hood, it is very simple and leverages the k_work api. an example of how it could be used:
```
int event1 = SAMPLE_SENSOR_A;
int event2 = SAMPLE_SENSOR_B;
int event3 = SAMPLE_SENSOR_c;

ZBUS_CHANNEL_PUBLISHER_DEFINE(pub1, &simple_chan, &event1);
ZBUS_CHANNEL_PUBLISHER_DEFINE(pub2, &simple_chan, &event2);
ZBUS_CHANNEL_PUBLISHER_DEFINE(pub2, &simple_chan, &event3);

void my_thread() {
    int msg;

   /* zbus_chan_publisher_start(struct zbus_publisher *publisher, k_timeout_t delay, k_timeout_t period) */
	zbus_chan_publisher_start(&pub1, K_NO_WAIT, K_MSEC(10));
	zbus_chan_publisher_start(&pub2, K_NO_WAIT, K_MINUTES(2));
	zbus_chan_publisher_start(&pub3, K_NO_WAIT, K_HOURS(1));

	while (1) {

		zbus_sub_wait_msg(zbus_obs, &chan, &msg, K_FOREVER);

		if (msg == SAMPLE_SENSOR_A) {
			/* take sensor A sample */
		} 
		if (msg == SAMPLE_SENSOR_B) {
			/* take sensor B sample */
		} 
		if (msg == SAMPLE_SENSOR_C) {
			/* take sensor C sample */
		} 	
	}
}
```
another way this can be leveraged is directly from a ISR...
```
int isr_msg = DATA_READY_MSG;
ZBUS_CHANNEL_PUBLISHER_DEFINE(isr_pub, &simple_chan, &isr_msg);

void my_isr( ... )
{
   /* process ISR */
   
   /* alert application */
   zbus_chan_publisher_reschedule(&isr_pub, K_NO_WAIT);
}

rodrigopex commented:
It's great! There are minor changes, but we will discuss them in the review phase. I suggest putting this feature in separate files (header and implementation) as it is entirely independent. The `ZBUS_WQ` should be in the zbus itself. I am going to use that in the `async listeners`.

rodrigopex commented:
The publisher is still interesting. The asynchronous listeners are related to reacting asynchronously. The publisher is about acting (pub) asynchronously and periodically. 

I usually use timers to publish to channels in a timely manner. I guess a publisher would be more suitable since that does not run in an ISR context. 

Maybe we need to think more.

brandon-exact committed:
zbus: add publishers to zbus

Adds the concept of publishers to async publish messages
to zbus channels. The new API is really just wrappers
around the work queue API.

A static variable of the publishers message must be provided
when delcaring the struct zbus_publisher with
ZBUS_CHANNEL_PUBLISHER_DEFINE

Signed-off-by: Brandon Allen <brandon.allen@exacttechnology.com>
